python/ray/data/_internal/execution/interfaces.py

```diff
@@ -237,6 +237,10 @@ class TaskContext:
     # an AllToAllOperator with an upstream MapOperator.
     upstream_map_transform_fn: Optional["MapTransformFn"] = None
 
+    # The Ray remote arguments of the fused upstream MapOperator.
+    # This should be set if upstream_map_transform_fn is set.
+    upstream_map_ray_remote_args: Dict[str, Any] = None
```

Comment:
Hmm, it's not ideal to pass this at runtime. Ideally, the optimizer would rewrite the downstream op's ray remote args to this value, instead of having each operator need to properly decide which of the two args to use and looking at the context.

Reply:
I agree with you. But currently it's hard to avoid this. For most operators, we are already doing the way you mentioned. 
 `upstream_map_ray_remote_args`, along with `upstream_map_transform_fn`, are used only for `RandomShuffle`. Because the corresponding `AllToAllOperator` physical op  itself doesn't directly do the shuffle. instead, it uses `ExchangeTaskScheduler` to launch new tasks to do the shuffle. That's why we need this ad-hoc handling here. I'll add a TODO here.